### PR TITLE
[Email] Update send email flow base on new design #606

### PR DESCRIPTION
**Pull request by DingmaomaoBJTU**
## Description

Update send email flow. Including:
Slot adjust by name;
Content confirmation;
Bypass slot by "skip";

## Related Issue
[Email] Update send email flow base on new design #606

## Checklist
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the appropriate unit tests
- [ ] I have updated related documentation

If this contains changes that needs to be replicated between the Enterprise Template <-> Virtual Assistant
- [ ] A duplicate issue is filed to track future  work.

If you have updated responses or `.lu` files:
- [ ] All languages have been updated
- [ ] You have tested deployment with your new models

### CODE REVIEW

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/Resources/SendEmailResponses.json [hunk 1]
```diff
@@ -11,21 +11,82 @@
...
     ],
     "inputHint": "expectingInput"
   },
   "NoMessageBody": {
     "replies": [
-
       {
-        "text": "What's the message?",
-        "speak": "What's the message?"
+        "text": "What do you want to add as the message?",
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/Resources/SendEmailResponses.json]: as the message body

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/Resources/SendEmailResponses.json [hunk 2]
```diff
@@ -11,21 +11,82 @@
...
+        "text": "Including a subject is often appreciated by contacts. Want to add a subject?",
+        "speak": "Including a subject is often appreciated by contacts. Want to add a subject?"
+      }
+    ],
+    "inputHint": "expectingInput"
+  },
+  "PlayBackMessage": {
+    "replies": [
+      {
+        "text": "I will playback the message for you to confirm:",
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/Resources/SendEmailResponses.json:41]: remove the last colon if nothing follows

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/Resources/SendEmailResponses.json [hunk 3]
```diff
@@ -11,21 +11,82 @@
...
+        "text": "Got it, this email will not be sent yet.",
+        "speak": "Got it, this email will not be sent yet."
+      }
+    ],
+    "inputHint": "expectingInput"
+  },
+  "GetRecreateInfo": {
+    "replies": [
+      {
+        "text": "Would you like to adjust the Participants, Subject or the Content?",
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/Resources/SendEmailResponses.json]: adjust -> change

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs [hunk 4]
```diff
@@ -80,33 +168,170 @@ public async Task<DialogTurnResult> CollectSubject(WaterfallStepContext sc, Canc
...
+                    var subjectInput = subject != null ? subject.ToString() : sc.Context.Activity.Text;
+
+                    if (!EmailCommonPhrase.GetIsSkip(subjectInput))
+                    {
+                        state.Subject = subjectInput;
+                    }
                 }
 
-                if (string.IsNullOrEmpty(state.Content))
+                if (state.Subject != null)
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs]: if (!string.IsNullOrEmpty(state.Subject))

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs [hunk 5]
```diff
@@ -80,33 +168,170 @@ public async Task<DialogTurnResult> CollectSubject(WaterfallStepContext sc, Canc
...
+            }
+        }
+
+        public async Task<DialogTurnResult> AfterUpdateSubject(WaterfallStepContext sc, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
+            {
+                var state = await EmailStateAccessor.GetAsync(sc.Context);
+
+                if (state.Subject != null)
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs]: if (!string.IsNullOrEmpty(state.Subject))

Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs]: actually if (!string.IsNullOrWhitespace(state.Subject))

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs (hunk 4 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs]: actually if (!string.IsNullOrWhitespace(state.Subject))

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs [hunk 6]
```diff
@@ -80,33 +168,170 @@ public async Task<DialogTurnResult> CollectSubject(WaterfallStepContext sc, Canc
...
+                return new DialogTurnResult(DialogTurnStatus.Cancelled, CommonUtil.DialogTurnResultCancelAllDialogs);
+            }
+        }
+
+        public async Task<DialogTurnResult> CollectText(WaterfallStepContext sc, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
+            {
+                var state = await EmailStateAccessor.GetAsync(sc.Context);
+                if (state.Content != null)
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/SendEmailDialog.cs]: if (!string.IsNullOrWhitespace(state.Content))

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Strings/EmailCommonStrings.zh.resx [hunk 7]
```diff
@@ -117,79 +117,100 @@
...
-    <value>空内容</value>
+  <data name="FromFormat" xml:space="preserve">
+    <value>发件人：{0}</value>
+  </data>
+  <data name="Cancel" xml:space="preserve">
+    <value>取消</value>
   </data>
   <data name="EmptySubject" xml:space="preserve">
-    <value>空标题</value>
+    <value>空主题</value>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/Shared/Resources/Strings/EmailCommonStrings.zh.resx]: i feel ‘空白主题’ sounds better :)

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/EmailSkill.csproj [hunk 8]
```diff
@@ -303,4 +306,8 @@
     <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/EmailSkill.csproj]: this is not needed

**Review by lzc850612 — CHANGES_REQUESTED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/Resources/SendEmailResponses.json (hunk 2 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/emailskill/Dialogs/SendEmail/Resources/SendEmailResponses.json:41]: There will be an adaptive card which display the message following this sentence. So keep the colon~

**Review by lzc850612 — APPROVED** [review]

**Pull request merged**